File: test/java/brs/web/api/http/handler/SetAliasTest.java
package brs.web.api.http.handler;

import brs.Attachment;
import brs.Blockchain;
import brs.Signum;
import brs.SignumException;
import brs.Constants;
import brs.common.QuickMocker;
import brs.common.QuickMocker.MockParam;
import brs.fluxcapacitor.FluxCapacitor;
import brs.fluxcapacitor.FluxValues;
import brs.services.AliasService;
import brs.services.ParameterService;
import brs.web.api.http.common.APITransactionManager;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import javax.servlet.http.HttpServletRequest;

import static brs.TransactionType.Messaging.ALIAS_ASSIGNMENT;
import static brs.web.api.http.common.JSONResponses.*;
import static brs.web.api.http.common.Parameters.ALIAS_NAME_PARAMETER;
import static brs.web.api.http.common.Parameters.ALIAS_URI_PARAMETER;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.powermock.api.mockito.PowerMockito.mockStatic;

@RunWith(PowerMockRunner.class)
@PrepareForTest(Signum.class)
public class SetAliasTest extends AbstractTransactionTest {

  private SetAlias t;

  private ParameterService parameterServiceMock;
  private Blockchain blockchainMock;
  private AliasService aliasServiceMock;
  private APITransactionManager apiTransactionManagerMock;

  @Before
  public void setUp() {
    mockStatic(Signum.class);

    parameterServiceMock = mock(ParameterService.class);
    blockchainMock = mock(Blockchain.class);
    aliasServiceMock = mock(AliasService.class);
    apiTransactionManagerMock = mock(APITransactionManager.class);

    FluxCapacitor mockFluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.PRE_POC2, FluxValues.DIGITAL_GOODS_STORE);
    when(Signum.getFluxCapacitor()).thenReturn(mockFluxCapacitor);

    t = new SetAlias(parameterServiceMock, blockchainMock, aliasServiceMock, apiTransactionManagerMock);
  }

  @Test
  public void processRequest() throws SignumException {
    final String aliasNameParameter = "aliasNameParameter";
    final String aliasUrl = "aliasUrl";

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ALIAS_NAME_PARAMETER, aliasNameParameter),
        new MockParam(ALIAS_URI_PARAMETER, aliasUrl)
    );

    mockStatic(Signum.class);
    final FluxCapacitor fluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE);
    when(Signum.getFluxCapacitor()).thenReturn(fluxCapacitor);
    doReturn(Constants.FEE_QUANT_SIP3).when(fluxCapacitor).getValue(eq(FluxValues.FEE_QUANT));

    final Attachment.MessagingAliasAssignment attachment = (Attachment.MessagingAliasAssignment) attachmentCreatedTransaction(() -> t.processRequest(req), apiTransactionManagerMock);
    assertNotNull(attachment);

    assertEquals(ALIAS_ASSIGNMENT, attachment.getTransactionType());
    assertEquals(aliasNameParameter, attachment.getAliasName());
    assertEquals(aliasUrl, attachment.getAliasURI());
  }

  @Test
  public void processRequest_missingAliasName() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ALIAS_NAME_PARAMETER, null),
        new MockParam(ALIAS_URI_PARAMETER, "aliasUrl")
    );

    assertEquals(MISSING_ALIAS_NAME, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectAliasLength_nameOnlySpaces() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ALIAS_NAME_PARAMETER, "  "),
        new MockParam(ALIAS_URI_PARAMETER, null)
    );

    assertEquals(INCORRECT_ALIAS_LENGTH, t.processRequest(req));
  }


  @Test
  public void processRequest_incorrectAliasLength_incorrectAliasName() throws SignumException {
    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ALIAS_NAME_PARAMETER, "[]"),
        new MockParam(ALIAS_URI_PARAMETER, null)
    );

    assertEquals(INCORRECT_ALIAS_NAME, t.processRequest(req));
  }

  @Test
  public void processRequest_incorrectUriLengthWhenOver1000Characters() throws SignumException {
    final StringBuilder uriOver1000Characters = new StringBuilder();

    for (int i = 0; i < 1001; i++) {
      uriOver1000Characters.append("a");
    }

    final HttpServletRequest req = QuickMocker.httpServletRequest(
        new MockParam(ALIAS_NAME_PARAMETER, "name"),
        new MockParam(ALIAS_URI_PARAMETER, uriOver1000Characters.toString())
    );

    assertEquals(INCORRECT_URI_LENGTH, t.processRequest(req));
  }

}

File: src/test/java/brs/http/SetAliasTest.kt
package brs.http

import brs.Attachment
import brs.Blockchain
import brs.DependencyProvider
import brs.common.QuickMocker
import brs.common.QuickMocker.MockParam
import brs.fluxcapacitor.FluxValues
import brs.http.JSONResponses.INCORRECT_ALIAS_LENGTH
import brs.http.JSONResponses.INCORRECT_ALIAS_NAME
import brs.http.JSONResponses.INCORRECT_URI_LENGTH
import brs.http.JSONResponses.MISSING_ALIAS_NAME
import brs.http.common.Parameters.ALIAS_NAME_PARAMETER
import brs.http.common.Parameters.ALIAS_URI_PARAMETER
import brs.services.AliasService
import brs.services.ParameterService
import brs.transaction.TransactionType
import brs.transaction.messaging.AliasAssignment
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class SetAliasTest : AbstractTransactionTest() {

    private var t: SetAlias? = null
    private lateinit var dp: DependencyProvider
    private var parameterServiceMock: ParameterService? = null
    private var blockchainMock: Blockchain? = null
    private var aliasServiceMock: AliasService? = null
    private var apiTransactionManagerMock: APITransactionManager? = null

    @Before
    fun setUp() {
        parameterServiceMock = mock()
        whenever(parameterServiceMock!!.getSenderAccount(any())).thenReturn(mock())
        blockchainMock = mock()
        aliasServiceMock = mock()
        apiTransactionManagerMock = mock()
        dp = QuickMocker.dependencyProvider(parameterServiceMock!!, blockchainMock!!, aliasServiceMock!!, apiTransactionManagerMock!!)
        t = SetAlias(dp)
    }

    @Test
    fun processRequest() {
        val aliasNameParameter = "aliasNameParameter"
        val aliasUrl = "aliasUrl"

        val request = QuickMocker.httpServletRequest(
                MockParam(ALIAS_NAME_PARAMETER, aliasNameParameter),
                MockParam(ALIAS_URI_PARAMETER, aliasUrl)
        )

        dp.fluxCapacitor = QuickMocker.fluxCapacitorEnabledFunctionalities(FluxValues.DIGITAL_GOODS_STORE)
        dp.transactionTypes = TransactionType.getTransactionTypes(dp)

        val attachment = attachmentCreatedTransaction({ t!!.processRequest(request) }, apiTransactionManagerMock!!) as Attachment.MessagingAliasAssignment
        assertNotNull(attachment)

        assertTrue(attachment.transactionType is AliasAssignment)
        assertEquals(aliasNameParameter, attachment.aliasName)
        assertEquals(aliasUrl, attachment.aliasURI)
    }

    @Test
    fun processRequest_missingAliasName() {
        val request = QuickMocker.httpServletRequest(
                MockParam(ALIAS_NAME_PARAMETER, null as String?),
                MockParam(ALIAS_URI_PARAMETER, "aliasUrl")
        )

        assertEquals(MISSING_ALIAS_NAME, t!!.processRequest(request))
    }

    @Test
    fun processRequest_incorrectAliasLength_nameOnlySpaces() {
        val request = QuickMocker.httpServletRequest(
                MockParam(ALIAS_NAME_PARAMETER, "  "),
                MockParam(ALIAS_URI_PARAMETER, null as String?)
        )

        assertEquals(INCORRECT_ALIAS_LENGTH, t!!.processRequest(request))
    }


    @Test
    fun processRequest_incorrectAliasLength_incorrectAliasName() {
        val request = QuickMocker.httpServletRequest(
                MockParam(ALIAS_NAME_PARAMETER, "[]"),
                MockParam(ALIAS_URI_PARAMETER, null as String?)
        )

        assertEquals(INCORRECT_ALIAS_NAME, t!!.processRequest(request))
    }

    @Test
    fun processRequest_incorrectUriLengthWhenOver1000Characters() {
        val uriOver1000Characters = StringBuilder()

        for (i in 0..1000) {
            uriOver1000Characters.append("a")
        }

        val request = QuickMocker.httpServletRequest(
                MockParam(ALIAS_NAME_PARAMETER, "name"),
                MockParam(ALIAS_URI_PARAMETER, uriOver1000Characters.toString())
        )

        assertEquals(INCORRECT_URI_LENGTH, t!!.processRequest(request))
    }

}
